python/tank/bootstrap/cached_configuration.py

```diff
@@ -0,0 +1,265 @@
+        """
+        Ensure that the configuration is up to date with the one
+        given by the associated descriptor.
+
+        This method fails gracefully and attempts to roll back to a
+        stable state on failure.
+        """
+        # make sure a scaffold is in place
+        self._config_writer.ensure_project_scaffold()
+
+        # stow away any previous versions of core and config folders
+        (config_backup_path, core_backup_path) = self._config_writer.move_to_backup()
+
+        # copy the configuration into place
+        try:
+            self._descriptor.copy(os.path.join(self._path.current_os, "config"))
+
+            # write out config files
+            self._config_writer.write_install_location_file()
+            self._config_writer.write_config_info_file(self._descriptor)
+            self._config_writer.write_shotgun_file()
+            self._config_writer.write_pipeline_config_file(
+                self._pipeline_config_id,
+                self._project_id,
+                self._entry_point,
+                self._bundle_cache_fallback_paths
+            )
+
+            # make sure roots file reflects current paths
+            self._config_writer.update_roots_file(self._descriptor)
+
+            # and lastly install core
+            self._config_writer.install_core(
+                self._descriptor,
+                self._bundle_cache_fallback_paths
+            )
+
+        except Exception, e:
+            log.exception("Failed to update configuration. Attempting Rollback. Error Traceback:")
+            # step 1 - clear core and config locations
+            log.debug("Cleaning out faulty config location...")
+            self._config_writer.move_to_backup()
+            # step 2 - recover previous core and backup
+            if config_backup_path is None or core_backup_path is None:
+                # there is nothing to restore!
+                log.error(
+                    "Irrecoverable error - failed to update config but no previous config to "
+                    "fall back on. Raising TankBootstrapError to abort bootstrap."
+                    )
+                raise TankBootstrapError("Configuration could not be installed: %s." % e)
+
+            else:
+                # ok to restore
+                log.debug("Restoring previous config %s" % config_backup_path)
+                filesystem.copy_folder(
+                    config_backup_path,
+                    os.path.join(self._path.current_os, "config")
+                )
+                log.debug("Previous config restore complete...")
+
+                log.debug("Restoring previous core %s" % core_backup_path)
+                filesystem.copy_folder(
+                    core_backup_path,
+                    os.path.join(self._path.current_os, "install", "core")
+                )
+                log.debug("Previous core restore complete...")
+
+        # @todo - prime caches (yaml, path cache)
```

Comment:
that's fine - it's purely an opportunity to improve performance now that we have the local bootstrap. We got a separate ticket for it, but it's lower prio!